### ISSUE TITLE
Better handling of back button and form saving to prevent losing state

### ISSUE THREAD

**Issue by keithamoss** [keithamoss/mapa#361]
- [ ] Save form state in local storage so the user can't lose it (consider how we handle hydration for old state / state saved on other devices)
- [x] Test back button behaviour on schema creation/editing from icon screens - it appears to go back to the schema list from the icon screens, not the actual previous screen.

**Comment by keithamoss** [comment]
Desktop-only issues, revisit later.

**Comment by keithamoss** [comment]
The behaviour is better now that symbols in schemas have their own unique URLs, but still breaks user expectations for layers below that (e.g. Choosing a symbol in SymbologyIconChooser or choosing a symbol style in SymbologyIconStyleChooser).

The ideal solution would be adding persistence of state to local storage for these components. Both the react-hook-form and the lower layers (e.g. SymbologyFieldEditor) that talk up to a form.

**Comment by keithamoss** [comment]
There's a few things at play here.

First, is us not using routes consistently in the deeper layers of the UI so that triggering the back button jumps the user much farther back than they were expecting. (This triggered the filing of this issue originally.)

Second, is that we're not persisting any state to local storage - forms, components, et cetera.

The second is a big piece of work, and I'm not currently sure it's worth the effort given the relatively small impact and the fact that adding deeper routes helps mitigate the issue.

Thirdly, how we deal with going back to entities that may have been deleted. e.g. Viewing a symbol, deleting it, and then hitting the back button.

...some time passes.

Well that was quite the saga!

Dug deep into approaches with (and without) react-router and came up blank. There's no way to do it consistently and reliably for the user without adding URL routes for every component, sub-component, sub-sub-component, et cetera.

The root of the issue is the browser's back button.

If it wasn't there (or we could disable it), and the user could only use the app's UI, we'd be fine.

The problem comes from a combination of deeply nested UI layers and our use of useState()

If we were persisting all draft changes to the redux store, AND if we were persisting those parts of the redux store to local storage, then we could deliver a 100% consistent experience where the page would respond 100% consistently no matter how the user landed on a URL.

The routes we'd have to add would look something like this list (partial list only):

```
path: 'SchemaManager/Edit',
path: 'SchemaManager/Edit/EditSymbol/1',
path: 'SchemaManager/Edit/EditSymbol/1/ChooseSymbol',

path: 'SchemaManager/Create',
path: 'SchemaManager/Create/AddNewSymbol',
path: 'SchemaManager/Create/AddNewSymbol/ChooseSymbol',
path: 'SchemaManager/Create/AddNewSymbol',

path: 'SchemaManager/Create',
path: 'SchemaManager/Create/EditSymbol/1',
path: 'SchemaManager/Create/EditSymbol/1/ChooseSymbol',

path: 'SchemaManager/Create',
path: 'SchemaManager/Create/DefaultSymbology',
path: 'SchemaManager/Create/DefaultSymbology/ChooseSymbol',

SchemaManager/Edit/69
SchemaManager/Edit/69/EditSchemaField/1
SchemaManager/Edit/69/EditSchemaField/1/EditSymbol
SchemaManager/Edit/69/EditSchemaField/1/EditSymbol/ChooseSymbol

SchemaManager/Edit/69
SchemaManager/Edit/69/AddSchemaField
SchemaManager/Edit/69/AddSchemaField/EditSymbol
SchemaManager/Edit/69/AddSchemaField/EditSymbol/ChooseSymbol
```

We tried various approaches that didn't involve adding all of these URL routes, including adding stating to the current URL (using the magic empty string trick)

`navigate('', { state: { actionTaken: 'onEditSymbol', symbolId: symbol.id } });`

And then using `useLocation() and usePrevious(useLocation())` or `window.addEventListener('popstate')` or `router.subscribe((state) => { } )` to listen for changes to work out what we had to do in response.

There's a relevant thread here talking about how react-router got rid of v5's useHistory(), but ultimately that wouldn't have solved it due to the below issues.

Ref: https://github.com/remix-run/react-router/issues/9422

e.g. The user hits back to navigate AWAY from the 'Choose an icon for this symbol screen' in SymbologyFieldEditor so in SymbologyFieldEditor we need to listen for that history POP event to know when to call the 'Close the icon choose dialog' function.

That gets tricky though, because history POP events only let you know what the current state should be, not where you came from. So you end up having to usePrevious() on the state to work out what action we have to do to update the UI to reflect the change.

e.g. The user hits back to navigate AWAY from the 'Choose an icon for this symbol screen' in SymbologyFieldEditor, so we need to make that UI update in SymbologyFieldEditor, but then in its parent component SchemaSymbologyManager we would also be listening to history POP events that would need to know to ignore that POP event because SymbologyFieldEditor is handling it.

Complexity abounds.

How could we have handled it better if we were starting again?

Well you'd think one option would be to give each full screen component its own react-router route?

Except that doesn't really work when you have sub-components (and sub-sub-components) that rely on being fed draft local state from their parent component.

e.g. Adding a group when editing a schema (its only stored in the parent component's useState()) and then navigating to the 'Edit symbol' route that's a separate component and has no knowledge of that group you just added because all it can do is pull from the schema stored in Redux from the main API calls.

Yes, sure, we could address THAT issue by persisting editable component state in either the redux store and/or useLocalStorage(), and pulling the list of groups from there when you navigate to the 'Edit symbol' component.

But I'm not about to rewrite all of the layers of UI from scratch to make that work given the impact of this issue is relatively low.

tl;dr Fuck it. We could make it work, but since the intent is that this is used as a PWA anyway, which has no back button, it's not worth the time.

We'll try the 'separate routes and use redux/local storage' approach next time we design a new UI screen that has at least two layers to it.

Here's some bits of code we hacked together to test some of these concepts, kept here for posterity:

```
export const addComponentToEndOfURLPathname = (pathname: string, actionName: string) => {
	let pathAddition = '';

	if (pathname.slice(-1) !== '/') {
		pathAddition = '/';
	}
	pathAddition = `${pathAddition}${actionName}`;

	return `${pathname}${pathAddition}`;
};

export const removeComponentFromEndOfURLPathname = (pathname: string, actionName: string) => {
	let pathnameCopy = pathname;

	// Trim any trailing slash
	if (pathnameCopy.slice(-1) === '/') {
		pathnameCopy = pathnameCopy.slice(0, -1);
	}

	if (pathnameCopy.includes(`/${actionName}`) === true) {
		pathnameCopy = pathnameCopy.replace(`/${actionName}`, '');
		return pathnameCopy;
	}

	// Just in case, keep the user on this page and show the Sentry report dialog
	Sentry.showReportDialog();
	return pathname;
};
```
